### DISCUSSION TITLE
Courant Number in Finite Volume Navier Stokes

### DISCUSSION THREAD

**Discussion by ABallisat** [idaholab/moose#24093] [Q&A Modules: Navier-Stokes] [answered]
Hello all,

I am trying to calculate the Courant number for a cell using INSFV within a custom kernel. I am trying to implement summing the Courant number over all faces which requires knowledge of the velocity at each face. I am using Rhie Chow interpolation for calculating the face velocities however I am getting an error. The code I am using is below

```cpp
  // Local courant number
  ADReal courantNumber = v.norm() *_face_info->faceArea();
  const ElemInfo currentElemInfo = ElemInfo(currentElement);

  for (int i=0; i<currentElement->n_sides(); ++i)
  {
    if (i != sideID) // because we already calculated it in the declaration
    {
      const FaceInfo * f = _mesh.faceInfo(currentElement, i);
      // FaceInfo objects are not created for every face on every element. They are only created once
      // per boundary between two elements and the element with the lower elementID (number) stores it.
      // If it is a null pointer then we need to get it from the neighbour
      if (f == nullptr)
      {
        const Elem * neighbourElement = currentElement->neighbor_ptr(i);
        const unsigned int faceIndex = currentElement->opposite_side(i);
        f = _mesh.faceInfo(neighbourElement, faceIndex);
      }
      ADRealVectorValue v = _rc_vel_provider.getVelocity(_velocity_interp_method, *f, _tid);
      courantNumber += v.norm() * f->faceArea();
    }
  }
  courantNumber *= _dt/currentElement->volume();
```

The comments hopefully explain most of the logic I have learned looking through the source code for the INSFV module. The error I am getting is
```
Attempted access into CellCenteredMapFunctor 'a' with a key that does not yet exist in the map. Make sure to fill your CellCenteredMapFunctor for all elements you will attempt to access later.
```
which I am guessing comes from the Rhie Chow interpolator not being able to access the information of the other element needed to calculate the velocity on the face. Is there a way to do this or a much simpler way of going about this? I wouldn't be surprised if I am missing something. Any help greatly appreciated!

Thanks in advance.

**Comment by GiudGiud** [comment]
Hello

When are you evaluating the courant number?
If you run this kernel before the RC interpolator has run, it s unable to provide the coefficients. 
Typically on INITIAL, that is a problem

Guillaume

**Reply by ABallisat** [reply]
It is in the middle of a kernel's computeQpResidual function. I have copied the INSFVScalarFieldAdvection and changed the computeQpResidual function so the execution time should be the same as that. Looking at the docs I cannot see any execute on options for any of the INSFV kernels so being honest I am not sure exactly when this gets called but it should be the same as INSFVScalarFieldAdvection.

**Reply by GiudGiud** [reply]
Kernels indeed do not have execute_on options. They are executed on every linear and nonlinear iterations. 

The NSFV momentum advection kernels are essentially run in two parts, first you compute their contribution to the Rhie Chow coefficients, then the contribution to the residual / jacobian.

**Comment by lindsayad** [comment]
Why do you need the Courant number in your kernel? If you are iterating over faces of elements attached to the current face, then yea you're definitely not guaranteed to have a filled Rhie-Chow coefficient map because you may be iterating beyond the current process domain's stencil

**Reply by ABallisat** [reply]
ok good to know, apologies this shows my lack of understanding. I am still investigating sharp interface advection and currently looking at the CICSAM scheme which uses the Courant number as part of the calculation of the the flux through faces. A decent approximation I suppose would probably be to calculate the value at the end of a timestep and then use that in the next step, can you foresee any problems with that? Would the coefficient map be filled by the end of the timestep?

**Reply by lindsayad** [reply]
If you're basing off of `INSFVScalarFieldAdvection` then is it enough to just compute the Courant number with the velocity from the current face (given by `_face_info`)?

**Reply by ABallisat** [reply]
The CICSAM scheme uses the cell Courant number, calculated as the sum of the Courant number at each face over all faces of a cell. I had considered using only the face value and am still working out the implications of making that approximation. Ideally I would like to do it with all the faces if possible. I guess the challenge then is trying to make sure that the Rhie Chow interpolator's coefficient map is fully populated, is there a way to check this or force it to populate before it calls the computeQpResidual of the kernel?

**Reply by lindsayad** [reply]
Is your kernel acting on a passive scalar, or what equation is it contributing to (may be multiple)?

**Reply by ABallisat** [reply]
I am solving an advection diffusion equation for a scalar and this kernel is being used in place of the INSFVScalarFieldAdvection kernel, hopefully as a drag and drop replacement. The Courant number is part of the calculation of the value of the scalar at the cell face, this is all to do with advecting sharp interfaces for multiphase flows. The advection velocity comes from the solution of the weakly compressible Navier Stokes equations hence why I am doing this using the finite volume framework.

**Reply by lindsayad** [reply]
Could you give me access to your code and input? We have several ideas for combatting your problem but I want to be 100% sure my hypothesis about your issue is correct

**Reply by ABallisat** [reply]
Thanks yes I should be able to, I will need to throw it into a separate repo but should hopefully get time to do that today.

**Comment by GiudGiud** [comment]
Are you using uniform refinement / adaptivity ?

What kind of boundary conditions do you have?

I ll add more info to this error message you are getting

**Reply by ABallisat** [reply]
I am not using any refinement or adaptivity, it is a fixed QUAD4 mesh built using a GeneratedMeshGenerator. My test case is a simple circle being advected in a constant diagonal velocity field in 2D. The inlet BCs are fixed INSFVInletVelocityBC of value [0.1, 0.1] on the left and bottom edges, the pressure is fixed at 0 on the right and top edges (the domain is a square).

**Reply by GiudGiud** [reply]
I wonder if we forgot an edge case in those combined boundaries.
Would you mind having the velocity inlet through a single boundary (and set the other one, for example left, as noslip or freeslip)
and the pressure outlet through a single boundary (and set the other one, for example right, as noslip or freeslip) ?

thank you!
This might not be it, I apologize in advance if this is not helpful

**Reply by ABallisat** [reply]
Honestly I'm all ears to suggestions so happy to give it a go. I will run it this afternoon and get back to you :) thanks for the help

**Reply by ABallisat** [reply]
Unfortunately removing the boundary conditions didn't work, I tried all combinations of removing velocity and pressure boundary conditions, and having none at all and the error still occured

**Reply by GiudGiud** [reply]
ok Alex has a pretty good idea of why it's happening. It s indeed not related to boundary conditions.

**Reply by lindsayad** [reply]
Yea I do have a strong guess, but I'd like to verify

**Reply by ABallisat** [reply]
Thanks both, I have just added you to the repo, let me know if you have any issues accessing it. Thanks again for your help!

**Reply by lindsayad** [reply]
Can you run your problem in debug mode? There are divide by zero errors due to `acceptorValue` and `upwindValue` being the same value. Please fix those so that I can get to the bad 'a' coefficient access. I've also opened https://github.com/ABallisat/CICSAM/pull/1 to update your code to the current MOOSE

**Reply by ABallisat** [reply]
Thanks, I will take a look at them and fix that issue. I know I am not going to get much time in the next week so I may be quiet till the end of the month but will be straight back onto this after that.

**Reply by lindsayad** [reply]
sounds good!

**Reply by ABallisat** [reply]
I've fixed the divide by zero error now and pushed up the code changes. On one processor it happily runs, I have done it for a couple of hundred time steps. It is overly compressive at the moment so the results aren't quite right yet but it you get a converged answer that looks reasonable. It immediately crashes if I try to run it with more than 1 processor with the same error about the key not existing in the CellCentredMapFunctor, hopefully you see the same behaviour.

**Reply by lindsayad** [reply]
@ABallisat can you create an issue on your repo and mention me on it? Then I can add it to my task list. GitHub projects seemingly can't handle discussions yet :-/

**Reply by ABallisat** [reply]
Done, let me know if you have any problems with it. I will keep digging into it this week on my end and see if I can spot anything. Thanks for your help with this!

**Reply by ABallisat** [reply]
I have found a (potential) fix - changing the _elem_range to go over active_semilocal_elements rather than active_local_subdomain_set_elements in the insfvSetup function of the INSFVRhieChowInterpolator class seems to fix it, that is changing the insfvSetup function to
```cpp
void
INSFVRhieChowInterpolator::insfvSetup()
{
  _elem_range =
      std::make_unique<ConstElemRange>(_mesh.active_semilocal_elements_begin(),
                                       _mesh.active_semilocal_elements_end());
}
```
The biggest issue I can foresee with this is that the elements are no longer constrained by the subdomain ID but I can't see any functionality in the docs for the mesh class to constrain semilocal elements by subdomain ID. Thoughts?

**Reply by GiudGiud** [reply]
You can check elem->subdomain_id inside the for loops using that range.

Extending the RC interpolator to semi-local elements should require more ghosting

**Reply by ABallisat** [reply]
Good to know thanks. I did also extend the ghosting (apologies, forgot to mention it). I added in the following to the validParams:
```cpp
  params.addRelationshipManager("ElementSideNeighborLayers",
                              Moose::RelationshipManagerType::GEOMETRIC,
                              [](const InputParameters &, InputParameters & rm_params)
                              { rm_params.set<unsigned short>("layers") = 3; });
```
Not sure if that is excessive amounts of ghosting but it works.

**Reply by GiudGiud** [reply]
I'm slightly concerned that a loop on semi-local elements will mess up the reduction of Rhie Chow coefficients. 
Basically, the contribution of each kernel to the coefficient would be double counted.

Could you please use the `FunctorAux` to output the a_x and a_y RC coefficients to auxiliary variables?
With and without the loop on semi-local, using MPI

If the coefficients are not the same then that's the problem. If they are the same then we're good

**Reply by lindsayad** [reply]
I'm not a big fan of this change. It will lead to more computation than is necessary

**Reply by GiudGiud** [reply]
I think it s ok to understand if the Courant-number-based interpolation scheme is good for the problem at hand.
If it is, then we can work on a less expensive change.

**Reply by lindsayad** [reply]
Yea I agree about that

**Reply by lindsayad** [reply]
But ignoring the expense, there will be the issue of correctness which you identified in that you will double count after doing the parallel communication

**Reply by ABallisat** [reply]
I have run it with and without the modification and they appear to be the same when run for both a constant velocity and variable velocity test cases (setting the bottom and left sides on a square to 0.1+sin(t)):

<img width="1393" alt="image" src="https://github.com/idaholab/moose/assets/40231544/97eb5142-4308-48bc-b283-011448e5b942">

Both the a_x and a_y coefficients give those plots. I have not run it in 3D but I expect the a_z coefficient will be the same. I think that is suggesting that it is not double counting if I am interpreting this correctly.

**Reply by GiudGiud** [reply]
Hello

Did you run this in parallel? We dont expect double counting in the serial case
How many ranks?

Guillaume

**Reply by ABallisat** [reply]
Yes it was run across 32 cores split across multiple nodes

**Reply by GiudGiud** [reply]
Interesting. Can you run a variable variation across the entire domain?
Also what do you mean by constant velocity? Are you not solving for velocity?

**Comment by ABallisat** [comment]
Sorry not sure I understand what you mean by a variable variation, could you clarify please?

I ran it with full fat INSFV so solving for velocity for two different cases. The first was a constant velocity regime, it is initialised to [0.1, 0.1] everywhere and the inlet is set be the same. In this way the INS equations are being solved but the velocity remains the same for all time. The second (variable) case is starting with an initial velocity of [0.1, 0.1] everywhere and then setting the inlet to [0.1+sin(t), 0.1+sin(t)] so that the solution of the INS equations leads to transient variations in the velocity field. In both cases I found the a coefficients to be the same, and both were run in parallel as above. Hope that makes sense!

**Reply by GiudGiud** [reply]
ok this all makes sense.
I meant a velocity field that depends on space, that is not constant everywhere.

Was there a non-zero pressure gradient in those simulations?

**Reply by ABallisat** [reply]
Yes, example from a time step below. There are inlets on the bottom and left sides of that square.

<img width="736" alt="image" src="https://github.com/idaholab/moose/assets/40231544/d7f3d457-0fb5-4f45-af4d-86e0b438d863">

**Reply by GiudGiud** [reply]
Ok I m inclined to believe we’re in the green. I was expecting an issue

**Reply by lindsayad** [reply]
There is probably no double counting because there are no elemental kernels active. That is the only thing `_elem_range` is used for

**Reply by lindsayad** [reply]
I would try reverting your `_elem_range` change and see whether just this mod that you made
```
  params.addRelationshipManager("ElementSideNeighborLayers",
                              Moose::RelationshipManagerType::GEOMETRIC,
                              [](const InputParameters &, InputParameters & rm_params)
                              { rm_params.set<unsigned short>("layers") = 3; });
```
is sufficient. I suspect that it is

**Reply by lindsayad** [reply]
Never mind. There is the time derivative term. I am shocked there is no exodiff between serial and parallel

**Reply by lindsayad** [reply]
If I make that change to `_elem_range` and run our test suite with `-p2` I see
```
test:finite_volume/materials/flow_diode.friction_flow_diode ................................. FAILED (EXODIFF)
test:finite_volume/wcns/boundary_conditions.dirichlet/using_velocity ........................ FAILED (EXODIFF)
test:finite_volume/pwcns/channel-flow.transient ............................................. FAILED (EXODIFF)
test:finite_volume/wcns/boundary_conditions.dirichlet/using_mdot ............................ FAILED (EXODIFF)
test:finite_volume/wcns/boundary_conditions.flux/direct ..................................... FAILED (EXODIFF)
test:finite_volume/wcns/channel-flow.transient .............................................. FAILED (EXODIFF)
test:finite_volume/ins/solidification.solidification_exo_pipe ............................... FAILED (EXODIFF)
test:finite_volume/pwcns/channel-flow.transient-action ...................................... FAILED (EXODIFF)
test:finite_volume/pwcns/boundary_conditions.from_mdot_action ............................... FAILED (EXODIFF)
test:finite_volume/ins/iks/flow-around-square.flow-around-square ............................ FAILED (EXODIFF)
test:finite_volume/ins/natural_convection.natural_circulation_dogleg ........................ FAILED (EXODIFF)
test:finite_volume/ins/solidification.solidification_centerline_profile ..................... FAILED (CSVDIFF)
test:finite_volume/wcns/materials.transient ................................................. FAILED (EXODIFF)
test:finite_volume/wcns/boundary_conditions.flux/from_velocity .............................. FAILED (EXODIFF)
test:finite_volume/wcns/channel-flow.transient-action ....................................... FAILED (EXODIFF)
test:finite_volume/wcns/natural_convection.natural_circulation_pipe ......................... FAILED (EXODIFF)
test:finite_volume/wcns/boundary_conditions.flux/from_velocity_action ....................... FAILED (EXODIFF)
test:finite_volume/wcns/boundary_conditions.flux/from_mdot .................................. FAILED (EXODIFF)
test:finite_volume/materials/flow_diode.controlled/time ..................................... FAILED (CSVDIFF)
test:finite_volume/wcns/boundary_conditions.flux/from_mdot_action ........................... FAILED (EXODIFF)
test:finite_volume/pins/channel-flow.friction ............................................... FAILED (EXODIFF)
test:finite_volume/wcns/channel-flow.turbulence ............................................. FAILED (EXODIFF)
test:finite_volume/pins/channel-flow.friction-action ........................................ FAILED (EXODIFF)
test:finite_volume/pins/channel-flow.friction-free-slip ..................................... FAILED (EXODIFF)
test:finite_volume/pins/channel-flow.friction-free-slip-action .............................. FAILED (EXODIFF)
test:finite_volume/materials/flow_diode.controlled/pdrop .................................... FAILED (CSVDIFF)
test:finite_volume/wcns/channel-flow.turbulence-action ...................................... FAILED (EXODIFF)
test:finite_volume/pins/channel-flow.friction-block ......................................... FAILED (EXODIFF)
test:finite_volume/pins/channel-flow.friction-block-action .................................. FAILED (EXODIFF)
test:finite_volume/ins/lid-driven.transient-with-temp ....................................... FAILED (EXODIFF)
test:finite_volume/pins/channel-flow.rz-by-parts ............................................ FAILED (EXODIFF)
test:finite_volume/materials/flow_diode.controlled/flow ..................................... FAILED (CSVDIFF)
test:finite_volume/ins/boussinesq.transient_wcnsfv .......................................... FAILED (EXODIFF)
test:finite_volume/ins/channel-flow.linear-friction ......................................... FAILED (EXODIFF)
test:finite_volume/ins/channel-flow.quadratic-friction ...................................... FAILED (EXODIFF)
test:finite_volume/pins/channel-flow/heated.transient ....................................... FAILED (EXODIFF)
test:finite_volume/ins/channel-flow.linear-friction-action .................................. FAILED (EXODIFF)
test:finite_volume/ins/channel-flow.quadratic-friction-action ............................... FAILED (EXODIFF)
test:finite_volume/pins/channel-flow/heated.transient-action ................................ FAILED (EXODIFF)
test:finite_volume/ins/channel-flow.discontinuous ........................................... FAILED (EXODIFF)
test:finite_volume/ins/channel-flow.transient ............................................... FAILED (EXODIFF)
test:finite_volume/ins/channel-flow.transient-action ........................................ FAILED (EXODIFF)
test:finite_volume/pins/mms/porosity_change.pressure-corrected ................... [min_cpus=8] FAILED (CRASH)
test:finite_volume/pins/mms/porosity_change.pressure-corrected-action ............ [min_cpus=8] FAILED (CRASH)
```

**Reply by lindsayad** [reply]
Ok I've opened https://github.com/idaholab/moose/pull/24454. I've tested that it works on the input file in your repo. I will create a PR to your repo once the MOOSE PR is merged ... or maybe I'll open it now

**Reply by lindsayad** [reply, marked as answer]
Opened https://github.com/ABallisat/CICSAM/pull/3

**Reply by ABallisat** [reply]
Thanks for doing that, I believe that part of it is now working. General update: the scheme works in MOOSE but needs some tuning, the original method as posed was designed for 1D and needs some modification to get the desired properties in 2D or 3D. I will report back if/when I get something working I am happy with. Thanks again for all your help with this!

**Reply by lindsayad** [reply]
You're welcome! Sorry it took a while!
